### PR TITLE
code updates to add a noise correlation factor between adjacent time slices for HB and HE channels

### PR DESCRIPTION
**Pull request by TaeunKwon**
#### PR description:

This PR updates MAHI to add a pedestal correlation factor between adjacent time slices to pedestal covariance matrix for HB and HE channels. A new class member is introduced in class `HBHEChannelInfo` to take and pass over the noise correlation factor to MAHI. This PR corresponds to [the previously merged PR](https://github.com/cms-sw/cmssw/pull/30999), which introduces the pedestal correlation factor to `HcalSiPMParametersRcd` tag for Run-3 MC scenario. The motivation and effect of changes by this PR are shown in [presentation by Taeun Kwon](https://indico.cern.ch/event/936151/contributions/3933432/attachments/2070567/3476700/pedestal.pdf). 

links of igprof report for CPU time measurements are given below. (Please see slide 15 and 19 of above presentation for the details)

igprof report for RECO without the pedestal correlation factor:
[https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO](https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO)

igprof report for RECO with the pedestal correlation factor:
[https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_NoiseCorrelation](https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_NoiseCorrelation)

igprof report for RECO without the pedestal correlation factor with PU:
[https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_wPU](https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_wPU)

igprof report for RECO with the pedestal correlation factor with PU:
[https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_NoiseCorrelation_wPU](https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_NoiseCorrelation_wPU)

links of igprof reports for DIGI are given below for the measurment of HLT timing cost change.
The sample used here : TTbar RelVal w/o PU

igprof report for DIGI without the pedestal correlation factor:
10.44 sec / 1000 evts taken for `mahiFit:minimization` module. 
https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_DIGI_woNC

igprof report for DIGI with the pedestal correlation factor:
11.25 sec / 1000 evts taken for `mahiFit:minimization` module. 
https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_DIGI_wNC

This PR is expected to affect on the HLT timing cost by 7~8% of time spent on  `mahiFit:minimization` module.

#### PR validation:

A basic technical test was performed: `runTheMatrix.py -l limited -i all --ibeos`

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This PR is not a backport and shouldn't be backported.

### CODE REVIEW

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 1]
```diff
@@ -78,10 +78,12 @@ void MahiFit::phase1Apply(const HBHEChannelInfo& channelData,
     if (iTS == nnlsWork_.tsOffset)
       tstrig += amplitude;
   }
-
+  const double noisecorr = channelData.noisecorr();
   tsTOT *= channelData.tsGain(0);
   tstrig *= channelData.tsGain(0);
 
+  std::cout << noisecorr << std::endl;
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc:85]: unprotected cout is verboten in production code.
Please remove.

Comment (TaeunKwon) [reply RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc]: I'm sorry, I forgot to remove the line before PR. The line was removed.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
from the slides p19 I see
- CPU time is measured using igprof
    - 97.34 sec / 1000 evts taken for mahiFit:minimization before the correction.
    - 107.61 sec / 1000 evts taken for mahiFit:minimization after the correction.

Please post a link to the igprof results in the PR description for completeness.

Does this affect the HLT timing cost as well?
If yes, what is the timing result there?

**Comment by TaeunKwon** [comment]
> from the slides p19 I see
> 
> * CPU time is measured using igprof
>   
>   * 97.34 sec / 1000 evts taken for mahiFit:minimization before the correction.
>   * 107.61 sec / 1000 evts taken for mahiFit:minimization after the correction.
> 
> Please post a link to the igprof results in the PR description for completeness.
> 
> Does this affect the HLT timing cost as well?
> If yes, what is the timing result there?

I added all the links of igprof reports I used in PR description and also ran the igprof on DIGI process to measure HLT timing cost before and after the pedestal noist correlation. It seems that This PR increases on the HLT timing cost by 2~3% of time spent on mahiFit:minimization module.

**Comment by slava77** [comment]
Looking at https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_DIGI/59
the igprof is dominated by the cout debug. Please update the profiler reports without the debugs.

Also, please clarify on the sample used in the HLT test, was it a sample with PU?

**Comment by slava77** [comment]
in a different way, RECO igprof is also distorted by the debugging module running MAHI
https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_wPU/644

Please also post the cmsDriver command used to make profiled setup.

**Comment by slava77** [comment]
The slides linked in the PR description are apparently covering only 2021 case.
Looking at just the number of hits in HB above some threshold, the number is much larger in the 2023 scenario. I suppose that this means more noise.

In 2021 (ttbar no-PU wf 11634, 10 events; red is with this PR)
all_OldVSNew_TTbar14TeV2021wf11634p0c_log10HBHERecHitsSorted_hbhereco__RECO_obj_obj_energy3

in 2023 (ttbar no-PU wf 12434, 10 events)
all_OldVSNew_TTbar14TeV2023wf12434p0c_log10HBHERecHitsSorted_hbhereco__RECO_obj_obj_energy3
Hits (mostly noise) are reconstructed with larger energy on the higher end of the spectrum.
Is this expected?

If there is analysis of 2023 setup available, it would be nice to have it added in the PR description, since this would be more relevant to understanding the noise treatment in the fit.

**Comment by TaeunKwon** [comment]
> in a different way, RECO igprof is also distorted by the debugging module running MAHI
> https://ktaeun.web.cern.ch/ktaeun/cgi-bin/igprof-navigator/igreport_RECO_wPU/644
> 
> Please also post the cmsDriver command used to make profiled setup.

Here is the cmsDriver command I used: 

For Gen-SIM,
` cmsDriver.py TTbar_14TeV_TuneCUETP8M1_cfi  --conditions auto:phase1_2024_realistic -n 10 --era Run3 --eventcontent FEVTDEBUG --relval 9000,100 -s GEN,SIM --datatier GEN-SIM --beamspot Run3RoundOptics25ns13TeVLowSigmaZ --geometry DB:Extended --fileout file:step1.root --python_filename step1_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2024_GenSimFull+DigiFull_2024+RecoFull_2024+HARVESTFull_2024+ALCAFull_2024.py --no_exec`

For DIGI, 
 `cmsDriver.py step2  --conditions auto:phase1_2024_realistic -s DIGI:pdigi_valid,L1,DIGI2RAW,HLT:@relval2021 --datatier GEN-SIM-DIGI-RAW -n 10 --geometry DB:Extended --era Run3 --eventcontent FEVTDEBUGHLT --filein  file:step1.root  --fileout file:step2.root --python_filename step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2024_GenSimFull+DigiFull_2024+RecoFull_2024+HARVESTFull_2024+ALCAFull_2024.py --no_exec`

For RECO, 
 `cmsDriver.py step3  --conditions auto:phase1_2024_realistic -n 10 --era Run3 --eventcontent RECOSIM,MINIAODSIM,DQM --runUnscheduled  -s RAW2DIGI,L1Reco,RECO,RECOSIM,EI,PAT,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM --datatier GEN-SIM-RECO,MINIAODSIM,DQMIO --geometry DB:Extended --filein  file:step2.root  --fileout file:step3.root --python_filename step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2024_GenSimFull+DigiFull_2024+RecoFull_2024+HARVESTFull_2024+ALCAFull_2024.py --no_exec`

As you pointed out, the analysis was done with ttbar no-PU wf 11634, 2021 case. If you think it will be better to have the analysis in 2023 case, I will do the analysis and add it in PR description.

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 2]
```diff
@@ -203,14 +203,20 @@ void MahiFit::doFit(std::array<float, 3>& correctedOutput, int nbx) const {
...
+const float MahiFit::minimize(const double noisecorr) const {
   nnlsWork_.invcovp.setZero(nnlsWork_.tsSize, nnlsWork_.nPulseTot);
   nnlsWork_.ampVec.setZero(nnlsWork_.nPulseTot);
 
   SampleMatrix invCovMat;
   invCovMat.setConstant(nnlsWork_.tsSize, nnlsWork_.tsSize, nnlsWork_.pedVal);
   invCovMat += nnlsWork_.noiseTerms.asDiagonal();
 
+  //Add off-Diagonal components up to first order
+  for (unsigned int i = 1; i < nnlsWork_.tsSize; ++i) {
```
[the remark this answers is not in the document]
Comment (TaeunKwon) [reply RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc]: Thanks! I added the conditional statement.

**Comment by slava77** [comment]
re https://github.com/cms-sw/cmssw/pull/31064#issuecomment-672073354

> Here is the cmsDriver command I used:

I see, this is for a somewhat extreme case with phase1_2024_realistic.
The cmsDriver commands do not show any PU details. Which PU was used for the HLT and reco-with-PU case?



> And for DIGI and RECO step, the code on below is inserted for 5C pedestal condition.

shouldn't the payloads be already in the GT phase1_2024_realistic? 
I'm just trying to understand if what's tested in the cms-bot tests with this PR is actually what's supposed to be running or if some elements are missing.

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 3]
```diff
@@ -203,14 +203,23 @@ void MahiFit::doFit(std::array<float, 3>& correctedOutput, int nbx) const {
...
+const float MahiFit::minimize(const double noisecorr) const {
   nnlsWork_.invcovp.setZero(nnlsWork_.tsSize, nnlsWork_.nPulseTot);
   nnlsWork_.ampVec.setZero(nnlsWork_.nPulseTot);
 
   SampleMatrix invCovMat;
   invCovMat.setConstant(nnlsWork_.tsSize, nnlsWork_.tsSize, nnlsWork_.pedVal);
   invCovMat += nnlsWork_.noiseTerms.asDiagonal();
 
+  //Add off-Diagonal components up to first order
+  if (noisecorr > 0.) {
```
Comment (slava77) [inline RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc:215]: on a second thought, shouldn't this be ```!= 0``` ?
I imagine anti-correlations are technically possible

**Comment by TaeunKwon** [comment]
For PU, the cmdDriver command is :

`cmsDriver.py TTbar_14TeV_TuneCUETP8M1_cfi  --conditions auto:phase1_2024_realistic -n 10 --era Run3 --eventcontent FEVTDEBUG --relval 9000,100 -s GEN,SIM --datatier GEN-SIM --beamspot Run3RoundOptics25ns13TeVLowSigmaZ --geometry DB:Extended --fileout file:step1.root  --python_filename step1_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2024PU_GenSimFull+DigiFullPU_2024PU+RecoFullPU_2024PU+HARVESTFullPU_2024PU+NanoFull_2024PU.py --no_exec`

`cmsDriver.py step2 --conditions auto:phase1_2024_realistic --pileup_input das:/RelValMinBias_14TeV/CMSSW_10_6_1-106X_mcRun3_2021_realistic_v1_rsb-v1/GEN-SIM -n 10 --era Run3 --eventcontent FEVTDEBUGHLT -s DIGI:pdigi_valid,L1,DIGI2RAW,HLT:@relval2021 --datatier GEN-SIM-DIGI-RAW --pileup AVE_35_BX_25ns --geometry DB:Extended --filein file:step1.root --fileout file:step2.root --python_filename step2_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2024PU_GenSimFull+DigiFullPU_2024PU+RecoFullPU_2024PU+HARVESTFullPU_2024PU+NanoFull_2024PU.py --no_exec`

`cmsDriver.py step3  --conditions auto:phase1_2024_realistic --pileup_input das:/RelValMinBias_14TeV/CMSSW_10_6_1-106X_mcRun3_2021_realistic_v1_rsb-v1/GEN-SIM -n 10 --era Run3 --eventcontent RECOSIM,MINIAODSIM,DQM --runUnscheduled  -s RAW2DIGI,L1Reco,RECO,RECOSIM,EI,PAT,VALIDATION:@standardValidation+@miniAODValidation,DQM:@standardDQM+@ExtraHLT+@miniAODDQM --datatier GEN-SIM-RECO,MINIAODSIM,DQMIO --pileup AVE_35_BX_25ns --geometry DB:Extended --filein  file:step2.root  --fileout file:step3.root  --python_filename step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2024PU_GenSimFull+DigiFullPU_2024PU+RecoFullPU_2024PU+HARVESTFullPU_2024PU+NanoFull_2024PU.py --no_exec`

PU input is `/RelValMinBias_14TeV/CMSSW_10_6_1-106X_mcRun3_2021_realistic_v1_rsb-v1/GEN-SIM`


> shouldn't the payloads be already in the GT phase1_2024_realistic?

Yes, you are right, I forgot the condtions updated in the GT of phase1_2024_realistic. We don't need the lines.

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc (hunk 3 above)
Comment (TaeunKwon) [reply RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc]: That makes more sense. Thanks!

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> ---> test runtestRecoEgammaPhotonIdentification had ERRORS

this is likely transient and unrelated 
```
Failed to open the file 'root://cms-xrd-global.cern.ch//store/relval/CMSSW_7_6_0_pre4/RelValH130GGgluonfusion_13/MINIAODSIM/PU25ns_76X_mcRun2_asymptotic_v1-v1/00000/14954967-E74F-E511-BEF2-0026189438EF.root'
```

I think that it can be ignored for this PR.

**Comment by abdoulline** [comment]
One small tech.comment about  GT for 2024:  

(1) aforementioned  (by Taeun in customization fragment ) "5C" conditions for Run3, including 2024 ones were added as default  (to auto:phase1_2024_realistic)  in https://github.com/cms-sw/cmssw/pull/30195  and appeared in 11_2_0_pre2.

(2) noise correlation factors (per se) have been introduced in the default conditions just recently (before this PR)  in
https://github.com/cms-sw/cmssw/pull/30999  and first appeared in  CMSSW_11_2_X_2020-08-03-2300.

So in the actual IBs both (1)+(2) are present by default in auto:phase1_2024_realistic

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
current status from my review side:
- it looks like the igprof links in the pr description were updated (a note to confirm it would be nice)
- an update to the study initially linked in the PR, but using 2023 or 2024 setup instead of 2021 (which has little noise) is expected
- in https://github.com/cms-sw/cmssw/pull/31064#issuecomment-672060143 the behavior of the fit in 2023 setup is unclear, perhaps it will become clear after the above item is clear
image).

**Comment by slava77** [comment]
here is a chi2 plot (from the bin-by-bin DQM outputs, after suppressing the large "overflow" bin)

2021 case (wf 11634)
wf11634_chi2_HB

2023 case (wf 12434), which is supposedly dominated by noise
wf12434_chi2_HB

The 2023 is more "self-consistent" fitting of the noise alone, but the behavior (increased chi2) seems to be the same as in the 2021 case, which was investigated in more detail in the DPG talk linked in the PR description.
 Was the correlation meant to reduce the chi2 (from first principles)? (I'm trying to understand if the sign in the correlation term is right)

**Comment by TaeunKwon** [comment]
Thanks for the clarification!

> it looks like the igprof links in the pr description were updated (a note to confirm it would be nice)

I'm sorry, I updated igprof report for DIGI and CPU time increased after the Noise correlation factor as expected. I will leave a note next time. 

> an update to the study initially linked in the PR, but using 2023 or 2024 setup instead of 2021 (which has little noise) is expected

The analysis in the [presentation](https://indico.cern.ch/event/936151/contributions/3933432/attachments/2070567/3476700/pedestal.pdf) was done in 2024 setup with conditions of phase1_2024_realistic GT to test extreme case with large noise. 

> in #31064 (comment) the behavior of the fit in 2023 setup is unclear, perhaps it will become clear after the above item is clear

I'm so sorry I think I missed your question in the comment. For the energy of hits, the increase of energy of Hits in low energy region is expected and we observed that it causes the increase of the number of Jets. But we expect that it will be fine because newly emerged Jets after the noise correlation factor has lower Pt, less than 20 GeV. Distributions of Jet Pt and nJets are in my slide. If you think it would be better to have more checks, please let me know.

**Comment by TaeunKwon** [comment]
> Was the correlation meant to reduce the chi2 (from first principles)? (I'm trying to understand if the sign in the correlation term is right)

No, the noise correlation factor increases the chi2 value of the fit. But as discussed in the DPG talk, chi2 value before the correlation was too small, considering chi2/ndof. So, the conclusion was the increase of chi2 value is okay because it makes the mean of chi2/ndof more closer to 1.

**Comment by slava77** [comment]
> > an update to the study initially linked in the PR, but using 2023 or 2024 setup instead of 2021 (which has little noise) is expected
> 
> The analysis in the [presentation](https://indico.cern.ch/event/936151/contributions/3933432/attachments/2070567/3476700/pedestal.pdf) was done in 2024 setup with conditions of phase1_2024_realistic GT to test extreme case with large noise.

Ah, sorry, I was "mislead" by 2021 in the introductory slide.
OK for this item then.

**Comment by slava77** [comment]
> > Was the correlation meant to reduce the chi2 (from first principles)? (I'm trying to understand if the sign in the correlation term is right)
> 
> No, the noise correlation factor increases the chi2 value of the fit. But as discussed in the DPG talk, chi2 value before the correlation was too small, considering chi2/ndof. So, the conclusion was the increase of chi2 value is okay because it makes the mean of chi2/ndof more closer to 1.

Thank you for clarifying. When I read the slides on this part, it just seemed a bit more observational than from first principles.
For some future follow up it may be useful to have simplified injection tests where the input signal is well understood (Gaussian noise, properly unbiased pulse shape).

**Comment by TaeunKwon** [comment]
> Thank you for clarifying. When I read the slides on this part, it just seemed a bit more observational than from first principles.
> For some future follow up it may be useful to have simplified injection tests where the input signal is well understood (Gaussian noise, properly unbiased pulse shape).

Thanks for the suggestion, we will include such tests in further updates.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31064	600bf92 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show differences only in Run3 workflows, starting from HBHE rechit variables, propagating downstream. The HBHE changes seen in the tests are in line with the studies shown in the DPG meeting (slides linked in the PR description). The igprof show that some increase in CPU time is expected, partly from the MAHI fit itself and partly from somewhat minor increase in energy of the reconstructed hits leading to possibly more work in the downstream modules. In reports for HLT and offline with PU the cost increase in calling MahiFit::phase1Apply is about 7%, which translates to about 0.1% of offline reco time.

**Comment by qliphy** [comment]
+1

**Comment by mariadalfonso** [comment]
As a note for everybody.
The correlation is not properly implemented.

It's calculated in minimumBias where the main noise contribution is on the diagonal is the dark current, while in the hard scattering the photo statistics term dominate.
Should be derived and applied in reference to the pedval term.

**Pull request merged**